### PR TITLE
Support bucket versionning

### PR DESCRIPTION
**Pull request by cottephi**

### CODE REVIEW

**Comment by cottephi** [comment]
closes #17

**Comment by oussjarrousse** [comment]
This looks very good at a first glance. I will return to it later tonight.
Thank you for the PR.

**Comment by oussjarrousse** [comment]
@cottephi enabling versionning should also have an effect on get_object. With versionning enabled, get_object() can take a version_id: https://min.io/docs/minio/linux/developers/python/API.html#get-object-bucket-name-object-name-offset-0-length-0-request-headers-none-ssec-none-version-id-none-extra-query-params-none
and on list_objects() with include_version=True: https://min.io/docs/minio/linux/developers/python/API.html#list-objects-bucket-name-prefix-none-recursive-false-start-after-none-include-user-meta-false-include-version-false-use-api-v1-false-use-url-encoding-type-true

Don't you think adding support for bucket versionning should also include support for object versions and get_object with version id, also list_objects() when include_versions=True

If your need is satisfied by adding the functions:
- set_bucket_versioning()
- get_bucket_versioning()
 
I can include that quickly in a minor release...

if you actually intend to add support for mocking keeping versions of objects 
- list_objects() with include_versions=True
- get_object() with specific version_id
that would be a bigger change.

**Comment by cottephi** [comment]
Good point, I'll add those too

File: tests/test_minio_mock.py [hunk 1]
```diff
@@ -125,7 +225,7 @@ def test_list_objects(minio_mock):
     client = Minio("http://local.host:9000")
 
     with pytest.raises(S3Error):
-        objects = client.list_objects("no-such-bucket")
+        _ = client.list_objects("no-such-bucket")
```
Comment (oussjarrousse) [inline tests/test_minio_mock.py:228]: That's cleaner 👍

File: pytest_minio_mock/plugin.py [hunk 2]
```diff
@@ -46,24 +49,32 @@ class MockMinioObject:
...
 
-    def __init__(self, object_name, data):
+    def __init__(self, object_name, data, version_id=None):
         """
         Initialize the MockMinioObject with a name and data.
 
         Args:
             object_name (str): The name of the object.
             data (bytes or io.BytesIO): The data the object contains.
+            version_id (str, optional): The version of the object.
```
Comment (oussjarrousse) [inline pytest_minio_mock/plugin.py:62]: Why allowing version_id=None?
Why not always have a version_id?
Even if versionning is disabled, giving the object a version_id should not be a problem.
that would simplify `get_object()` and `put_object()` the case when buckets already had objects before versionning is ENABLED (by accessing [bucket_name][object_name] last version `...[object_name].keys())[-1]` if versionning is not ENABLED or if version_id was not provided, and [bucket_name][object_name][version_id] if versionning is enabled)

What do you think?

File: pytest_minio_mock/plugin.py [hunk 3]
```diff
@@ -322,11 +333,35 @@ def get_object(
...
         the_object = self.buckets[bucket_name][object_name]
+        if (
+                self.get_bucket_versioning(bucket_name).status == ENABLED
+                and isinstance(the_object, dict)
+        ):
+            # versioning is activated on the bucket, and the object is versioned
+            if not version_id:
+                version_id = list(the_object.keys())[-1]
+            try:
+                # TODO: should add version UUID validation, maybe with pydantic ?
```
Comment (oussjarrousse) [inline pytest_minio_mock/plugin.py:344]: `validators` is used elsewhere in this package. For `uuid` check https://python-validators.github.io/validators/reference/uuid/

File: pytest_minio_mock/plugin.py [hunk 4]
```diff
@@ -446,9 +481,24 @@ def put_object(
                 bucket_name=bucket_name,
                 object_name=None,
             )
-        self.buckets[bucket_name][object_name] = MockMinioObject(
-            object_name=object_name, data=data
-        )
+        if self.get_bucket_versioning(bucket_name).status == ENABLED:
```
Comment (oussjarrousse) [inline pytest_minio_mock/plugin.py:484]: I would do it this way to simplify the code:
        
        version = str(uuid4())
        obj = MockMinioObject(object_name=object_name, data=data, version_id=version)
        if self.get_bucket_versioning(bucket_name).status != DISABLED:
            self.buckets[bucket_name][object_name] = {}   # delete other versions
        self.buckets[bucket_name][object_name][version] = obj

That simplifies the code significantly... Similar changes need to happen in get_object()

File: pytest_minio_mock/plugin.py [hunk 5]
```diff
@@ -322,11 +333,35 @@ def get_object(
         """
         self._health_check()
         the_object = self.buckets[bucket_name][object_name]
+        if (
```
Comment (oussjarrousse) [inline pytest_minio_mock/plugin.py:336]: If a dictionary is always used as I suggested in my review for `put_object()` where a version_id is always used. then get_object can be simplified by doing the following:

    object_versions = self.buckets[bucket_name][object_name]
    
    # we try to get the version_id
    if  self.get_bucket_versioning(bucket_name).status != ENABLED
        version_id = list(the_object.keys())[-1]
    else:  # versionning is enabled
        if not version_id:
            version_id = list(the_object.keys())[-1]
            
    # we get the object using the version we got above
    try:
        the_object = object_versions[version_id]    
    except KeyError:
                raise S3Error(
                    message="The specified version does not exist",
                    resource=version_id,
                    request_id=None,
                    host_id=None,
                    response="mocked_response",
                    code=404,
                    bucket_name=bucket_name,
                    object_name=object_name,
                )

**Review by oussjarrousse — CHANGES_REQUESTED** [review]
Thank you for the PR. 

It is very substantial and adds missing functionality to the package.

I am requesting changes mainly to simplify the code by **always** storing objects in a dictionary of the form `{version_id: the_object}` in the `[bucket_name][object_name]`.

I suggested code changes in put_object() and get_object(), towards that goal. Changes elsewhere in the code might still be necessary.

Thank you again. I am looking forward for your reply.
Oussama

File: pytest_minio_mock/plugin.py (hunk 2 above)
Comment (cottephi) [reply pytest_minio_mock/plugin.py:62]: Indeed, it is much more simple. I'll do that.

File: pytest_minio_mock/plugin.py (hunk 3 above)
Comment (cottephi) [reply pytest_minio_mock/plugin.py]: done

**Comment by cottephi** [comment]
So, I ended up also adding `stat_object`, the `last_modified` key in Objects, and I moved all the S3Error to a new file, to clean up the code a bit.

**Comment by oussjarrousse** [comment]
Yes let's split it please, and let's just focus on adding Versionning. 
Can you please restrict the changes in this PR code directly related to versionning?
I appreciate the effort, but I want to the process clean.
Thank you

**Comment by cottephi** [comment]
No problem, I completely understand. I'll revert some commits and create a new branch tomorrow, and create a new PR once this one is merged.

**Comment by cottephi** [comment]
So, I rolled back this branch three commits, so now only bucket versioning remains. I made a new branch for the other enhancements, and I'll open a new PR once this one is merged

**Comment by oussjarrousse** [comment]
Ok, but for some reason I can **still** see all the changes of all the commits you made in this pull request.

**Comment by cottephi** [comment]
Really ? There should be only two commits, do you see more ? And do you see the new method def stat_object ?

**Comment by oussjarrousse** [comment]
I see 2 commits. But I still see changes to conftest.py, and the tests/*.py  that introduces breaking changes.

**Comment by oussjarrousse** [comment]
Here is what I will do, 
I will merge the pull request to a separate branch. 
Will edit it and merge it with main and let you know.

**Pull request merged**